interfaces/builtin/shared_memory.go

```diff
@@ -0,0 +1,162 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/snap"
+)
+
+const sharedMemorySummary = `allows two snaps to use predefined shared memory objects`
+
+const sharedMemoryBaseDeclarationSlots = `
+  shared-memory:
+    allow-installation:
+      slot-snap-type:
+        - app
+        - gadget
+    allow-connection:
+      plug-attributes:
+        shared-memory: $SLOT(shared-memory)
+    deny-auto-connection: true
+`
+
+// sharedMemoryInterface allows sharing sharedMemory between snaps
+type sharedMemoryInterface struct{}
+
+func (iface *sharedMemoryInterface) Name() string {
+	return "shared-memory"
+}
+
+func (iface *sharedMemoryInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              sharedMemorySummary,
+		BaseDeclarationSlots: sharedMemoryBaseDeclarationSlots,
+		AffectsPlugOnRefresh: true,
+	}
+}
+
+func (iface *sharedMemoryInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	sharedMemoryAttr, isSet := slot.Attrs["shared-memory"]
+	sharedMemory, ok := sharedMemoryAttr.(string)
+	if isSet && !ok {
+		return fmt.Errorf(`shared-memory "shared-memory" attribute must be a string, not %v`,
+			slot.Attrs["shared-memory"])
+	}
+	if sharedMemory == "" {
+		if slot.Attrs == nil {
+			slot.Attrs = make(map[string]interface{})
+		}
+		// shared-memory defaults to "slot" name if unspecified
+		slot.Attrs["shared-memory"] = slot.Name
+	}
+
+	pathError := func(value interface{}) error {
+		return fmt.Errorf(`shared-memory "write" attribute must be a list of strings, not %v`, value)
+	}
+	// check that we have a write path
+	writeAttr, ok := slot.Attrs["write"]
```

Comment:
so reading the spec and thinking about this more, I'm thinking about how we can support the read/write aspect of shared-memory to match content and system-files. 

For background, with the content interface, snaps can setup one way communication where the directory shared is writable from the snap providing the slot just by choosing a directory like $SNAP_DATA as the specifier in `read`, and then the plug side of connection only has rules allowing read access to the files shared. The slot can setup bi-directional sharing by choosing a directory that the slotting snap can write to as the source in the `write` setting, and then the plug side is allowed to write to that directory. 

However, we can't do the same thing here for shared-memory, since if we used the same pattern and had something like this:

```
slots:
  foo-mem:
    interface: shared-memory
    read:
      - /dev/shm/org.foo.mem
```

Then we would setup read-only permissions for the plug, but then what permissions do we setup for the slot side? We should arguably always setup write permissions to enable the one way communication from slot -> plug.

So given that, I think that what we should do is use the slot side settings for read/write to indicate the permissions for the plug side, but always specify write access for the slot.